### ISSUE TITLE
helpers.Page.pager cannot be configured to add accessibility spans

### ISSUE THREAD

**Issue by KrzysztofMadejski** [ckan/ckan#3104]
### CKAN Version if known (or site URL)

2.3.3
### Please describe the expected behaviour

Site should be WCAG accessible
### Please describe the actual behaviour

Cannot modify style of the `helpers.Page.pager` so that proper accessibility `span`s can be introduced
### Proposed solution

Introduce wcag_hide css class:

```
.wcag_hide { 
   position:absolute; 
   top:0; 
   left:-10000px; 
   width:1px; 
   height:1px; 
}

.wcag_hide2 {
    clip: rect(1px, 1px, 1px, 1px);
    display: block;
    position: absolute;
}

```

Modify `ckan.lib.helpers.Page` adding `page_format` to allow styling.
Optionally: modify default symbols

```
class Page(paginate.Page):
    # Curry the pager method of the webhelpers.paginate.Page class, so we have
    # our custom layout set as default.

    def pager(self, *args, **kwargs):
        kwargs.update(
            format=u"<div class='pagination pagination-centered'><ul>$link_previous ~2~ $link_next</ul></div>",
            symbol_previous=literal('<span class="wcag_hide">' + _('Previous page of results') + '</span>' + u'«'),
            symbol_next=literal('<span class="wcag_hide">' + _('Next page of results') + '</span>' + u'»'),
            page_format=literal(_('<span class="wcag_hide">Page </span>$pagenum')),
            curpage_attr={'class': 'active'}, link_attr={}
        )
        self.page_format = kwargs['page_format']
        del kwargs['page_format']

        return super(Page, self).pager(*args, **kwargs)

    # Put each page link into a <li> (for Bootstrap to style it)

    def _pagerlink(self, page, text, extra_attributes=None):
        if re.match(r'^(\d+)$', text):
            text = Template(self.page_format).safe_substitute({'pagenum': text})

        anchor = super(Page, self)._pagerlink(page, text)
        extra_attributes = extra_attributes or {}
        return HTML.li(anchor, **extra_attributes)
```

**Comment by wardi** [comment]
If this helper used a snippet instead of literal html it would be easy to override

**Comment by KrzysztofMadejski** [comment]
In the end it's dependent on `webhelpers.paginate` and I believe it should be resolved there -> https://bitbucket.org/krzysztofmadejski/webhelpers/issues/1/allow-templating-of-page-number-shown-so

I think `literal`is ok. It's recommended by WebHelpers when one want to embed html.

**Comment by wardi** [comment]
Does webhelpers.paginate give us something that we couldn't do with a pager template snippet that has a few if statements and a for loop?

If not, I'd prefer to drop the dependency and gain the easy overriding we get by using a snippet.

**Comment by KrzysztofMadejski** [comment]
That's an option. One (and only I believe) "complex" thing that paginate do is the `~3~` config option saying show me 3 pages left and right of the current one.

**Comment by wardi** [comment]
@KrzysztofMadejski do you have time to work on this?

The approach I'd like is to replace calls to `h.Page.pager` in our templates with a call to a snippet that renders a pager. This snippet might use a simpler helper for producing lists of adjacent pages on the left and right, but all the rendered HTML should be part of the snippet. The snippet would have blocks where they make sense for overridding the look and style of each link and span.

We can maintain compatibility with existing templates by making the `h.Page.pager` helper render the same snippet. How does that sound to you?

**Comment by KrzysztofMadejski** [comment]
I agree that would be better to have full templates in CKAN. I don't have
time now, but possibly will have soon. If so I will contact you where would
it be best to put our work in (this or other issues).

On Tuesday, 14 June 2016, Ian Ward <notifications@github.com> wrote:

> @KrzysztofMadejski <https://github.com/KrzysztofMadejski> do you have
> time to work on this?
>
> The approach I'd like is to replace calls to h.Page.pager in our
> templates with a call to a snippet that renders a pager. This snippet might
> use a simpler helper for producing lists of adjacent pages on the left and
> right, but all the rendered HTML should be part of the snippet. The snippet
> would have blocks where they make sense for overridding the look and style
> of each link and span.
>
> We can maintain compatibility with existing templates by making the
> h.Page.pager helper render the same snippet. How does that sound to you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ckan/ckan/issues/3104#issuecomment-225871993>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAz4N647RYHJas9MUMMSNa4Y5q1enIEmks5qLqSpgaJpZM4IzApr>
> .
>


-- 
*Krzysztof Madejski*
<karolina.wysocka@epf.org.pl>Code for Poland Program Manager / Koordynator
Koduj Dla Polski
krzysztof.madejski@epf.org.pl / +48 664 082 823

* <http://epf.org.pl/>*

**Comment by wardi** [comment]
I'll leave this issue here with our discussion so anyone that does work on it has a place to start.

**Comment by gleb-rudenko** [comment]
@wardi , I can start working on it.

**Comment by kmbn** [comment]
We decided to close old issues that are not actively worked on so that we can focus our effort and attention on issues affecting the current versions of CKAN.

If this issue is still affecting the version of CKAN you're working with now, please feel free to comment or reopen the issue.

If you do reopen this issue, please update it with new details. One reason it might not have been resolved in the past is that it wasn't clear how a contributor could address the issue.

**Comment by J-bytes** [comment]
I reopen this issue as it seems to still be a problem on ckan 2.8.6. The impossibility to easily change helpers.Page.pager to add accessibility spans remains a problem.

**Comment by mgifford** [comment]
This is what we did in Drupal:
https://www.drupal.org/docs/accessibility/hide-content-properly

We based on the [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate/blob/a00d0365c2e293979b6ebe5bf0c80c34dc913284/dist/css/style.css) in Drupal 8, although we had established a similar convention with an element-invisible class in Drupal 7. 

This is really a pretty simple convention.  It is also one that is reasonably well known with folks using popular CMS frameworks.
